### DISCUSSION TITLE
Increase stack size to 8192 bytes

### DISCUSSION THREAD

**Discussion by lmarmisa** [esp8266/Arduino#8652] [Q&A] [answered]
I have developed a tiny but powerful script language for the ESP8266 (supporting grammars and bult-in functions for int, double, unsigned, string and many other features). This little shell program works great. I use Arduino IDE and release 3.0.2.

I am facing a problem related to the stack size. The standard size of the stack (4096 bytes) limits the power of the shell program. 

:stack_top         :0x3fffefc0
:current_sp        :0x3ffffdc0
:stack_bottom      :0x3fffffbf
:stack_size_bytes  :4096
:loop_free_stack   :4032
:current_free_stack:3584
:unused_stack      :1024
:stack_overflow    :0

My design is very optimized in terms of memory, but parsers and lexical analyzers in general make massive use of function calls and even frequently use recursion. So I would like to know if there is a way for increasing the dimension of the stack. I think that 8192 bytes would allow very decent levels of recursion for my program.

I would have no problem to modify some parameters or files in the .arduino15 structure in order to define the new stack size, but I need some help in order to locate the place where I have to modify the source files (maybe a #define would be enough).

Thanks in advance for your help,

Luis

**Comment by earlephilhower** [comment, marked as answer]
-edit, wrong core! -

What I do for BearSSL which needs 6K+ of stack, is to just allocate a per-module stack and use stack thunking to swap stacks on-the-fly.  Regular code runs normally. but when you call the SSL commands (or, say, your global parser routine) the SP is set to a new value in DRAM, the routine is called, and then the SP is reset back to the original value.

See https://github.com/esp8266/Arduino/blob/master/cores/esp8266/StackThunk.cpp and https://github.com/esp8266/Arduino/blob/master/cores/esp8266/StackThunk.h

**Reply by lmarmisa** [reply]
Thanks a lot for you answer. Very interesting.

I need a little help with the macro make_stack_thunk(fcnToThunk).

I want to create a thunk with a function like this:

`boolean parse_i(char *input, int *val, char **error);`

I tried this:

```
...
#include "parseri.h"
...
extern "C" {
  boolean thunk_parse_i(char *input, int *val, char **error);
  make_stack_thunk(parse_i);
}
...
```

The Arduino IDE compiled the file cpp but a undefined reference to parse_i is detected in function thunk_parse_i (.text+0xc)(linking time).

I would appreciate if you could guide me to solve this problem.

**Reply by lmarmisa** [reply]
I have found the solution.

File "parseri.h":
```
...
extern "C" {
  boolean parse_i(char *input, int *val, char **error);
  boolean thunk_parse_i(char *input, int *val, char **error);
}
...
```
File "parseri.cpp":
```
...
#include "parseri.h"
...
//parse_i
boolean parse_i(char *input, int *val, char **error) {
  ...
}
//thunk_parse_i
make_stack_thunk(parse_i);
...
```
The functions stack_thunk_add_ref() and stack_thunk_del_ref() are used too before and after a call to  thunk_parse_i.

I think that a stack thunk of 6200 bytes is a bit limited for my needs. I will try to adapt the program StackThunk.cpp to my needs developing a new program based on yours.

Thank you.

**Reply by earlephilhower** [reply]
Nice work!  You could also remove the `extern "C" {}` section in both spots.  That just sets the function name to not use C++ name mangling.  It is present in the BearSSL thinks because BearSSL only wants C-names, but for your own code they can both be name mangled.

**Reply by lmarmisa** [reply]
If I remove the extern "C" directive in any of the two function headers, the undefined references come back again.

Apparently assembler code needs C-names during the link-time.
